Cookie bar cleanup

torotil opened the pull request.

torotil committed:
[google_analytics] Always ensure ga() exists.

Until the tracking script is loaded ga() is defined as dummy by the
usual GA snippets. We can use the same dummy to define it even when no
snippet is loaded (eg. cookies are being blocked by a cookie bar). This
way all tracking info is queued until the script is loaded (or not).

torotil committed:
[fb_pixel] Load script using campaignion_tracking_snippets().

torotil committed:
[fb_pixel] Use the variable module and services container.

torotil committed:
[google_analytics] Improve payment loading in hook_page_build().

torotil committed:
[ga] Provide a variable for embedding the snippet.

torotil committed:
[tracking] Support for multiple tracking events.

- Snippets can now define to which event they listen.
- Support for triggering events early and provide a simple way to
  trigger scripts even if the event has already been fired.

katzmo commented on campaignion_google_analytics/campaignion_google_analytics.module:
```diff
@@ -228,3 +228,21 @@ function campaignion_google_analytics_page_build(&$page) {
+(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
+m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
+ga('create', '$ga_id', 'auto');
```
`$ga_id` should probably not be quoted :wink:

torotil replied:
The quotes are needed in this version (otherwise it would be interpreted as a JS variable, or worse executed as code), but it’s probably much better to JSON encode it instead than to hard-code the quotes :thinking:

katzmo replied:
Right, the heredoc is so long I forgot it's there :see_no_evil:

Not sure JSON encoding is worth the trouble? Maybe putting it in curly braces would make the variable easier to recognize? :thinking:

torotil replied:
I did both :-)

torotil committed:
[ga] Use drupal_json_encode() to safely encode the variable for JS.

torotil committed:
[tracking] Facebook pixel & Google analytics now depend on tracking.

Adds dependencies and migrations for campaignion_google_analytics and
campaignion_facebook_pixel to depend on campaignion_tracking.

They rely on it to call their snippets.

torotil committed:
[tracking] Allow snippets to don’t wait for any event.

Passing an empty $snippet['event'] means that the snippet is executed
immediately.

torotil committed:
[tracking] Invoke alter-hook on the normalized snippets.

torotil committed:
[tracking] Make default tracking event configurable.

Introduces a new variable campaignion_tracking_default_event and adds
the variable module as dependency.

torotil committed:
[tracking] Don’t wait for a JS event before loading scripts by default.

This is needed for backwards compatibility. Currently the modules don’t
wait for an event without further configuration.

katzmo commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -184,13 +184,13 @@ function campaignion_tracking_webform_redirect_alter($redirect, Submission $subm
     }
     $snippet += ['event' => 'tracking-allowed'];
   }
+  drupal_alter('campaignion_tracking_snippets', $snippets);
```
Is that a benefit? The difference I see is that other modules now need to provide their additional snippets as array, including an event name, while before just the snippet would have been enough and the event name optional. The old version seemed safer to me. Am I overlooking anything else here? :thinking:

torotil replied:
Jep. The original hook is called before the normalization. So if you just *provide* snippets you can rely on the normalization. If on the other hand you want to manipulate snippets defined elsewhere you can rely on a canonical format, because the alter hook is called after the normalization.

torotil replied:
I mean:

* If you want to define snippets: Use `hook_campaignion_tracking_snippets()` and rely on the normalization.
* If you want to change snippets: Use `hook_campaignion_tracking_snippets_alter()` and rely on getting the normalized snippets as input for altering.

katzmo replied:
Makes sense :+1:
